rust/sbp/src/wire_format.rs

```diff
@@ -0,0 +1,210 @@
+//! Functionality for parsing/writing SBP message payloads with in memory buffers
+
+use std::{
+    self,
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+use bytes::{Buf, BufMut, BytesMut};
+
+pub trait WireFormat: Sized {
+    /// Minimum number of bytes this type will take in the frame.
+    const MIN_ENCODED_LEN: usize = mem::size_of::<Self>();
+
+    /// The actual number of bytes the type will take in the frame.
+    fn encoded_len(&self) -> usize {
+        Self::MIN_ENCODED_LEN
+    }
+
+    /// Write the type to a buffer.
+    fn write(&self, buf: &mut BytesMut);
+
+    /// Read the type out of a buffer, without checking any invariants.
+    fn parse_unchecked(buf: &mut BytesMut) -> Self;
+
+    /// Read the type out of a buffer.
+    fn parse(buf: &mut BytesMut) -> Result<Self, PayloadParseError> {
+        if buf.remaining() >= Self::MIN_ENCODED_LEN {
+            Ok(Self::parse_unchecked(buf))
+        } else {
+            Err(PayloadParseError {})
+        }
+    }
+}
+
+impl<T> WireFormat for Vec<T>
+where
+    T: WireFormat,
+{
+    // A variable length array with no elements takes 0 bytes
+    const MIN_ENCODED_LEN: usize = 0;
+
+    fn encoded_len(&self) -> usize {
+        self.iter().map(WireFormat::encoded_len).sum()
+    }
+
+    fn write(&self, buf: &mut BytesMut) {
+        for item in self.iter() {
+            item.write(buf);
+        }
+    }
+
+    fn parse_unchecked(buf: &mut BytesMut) -> Self {
+        let mut v = Vec::new();
+        while buf.remaining() >= T::MIN_ENCODED_LEN {
+            v.push(T::parse_unchecked(buf));
+        }
+        v
+    }
+}
+
+impl<T, const LEN: usize> WireFormat for [T; LEN]
+where
+    T: WireFormat,
+{
+    const MIN_ENCODED_LEN: usize = T::MIN_ENCODED_LEN * LEN;
+
+    fn write(&self, buf: &mut BytesMut) {
+        for item in self {
+            item.write(buf);
+        }
+    }
+
+    fn parse_unchecked(buf: &mut BytesMut) -> Self {
```

Comment:
Probably a premature optimization. But, this is pretty much the example right out of the docs for `MaybeUninit` so I don't feel too bad about this touch of unsafe